gnovm/cmd/gno/doc.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"path/filepath"
+
+	"github.com/gnolang/gno/gnovm/pkg/doc"
+	"github.com/gnolang/gno/tm2/pkg/commands"
+)
+
+type docCfg struct {
+	all        bool
+	src        bool
+	unexported bool
+	rootDirStruct
+}
+
+func newDocCmd(io *commands.IO) *commands.Command {
+	c := &docCfg{}
+	return commands.NewCommand(
+		commands.Metadata{
+			Name:       "doc",
+			ShortUsage: "doc [flags] <pkgsym>",
+			ShortHelp:  "get documentation for the specified package or symbol (type, function, method, or variable/constant).",
+		},
+		c,
+		func(_ context.Context, args []string) error {
+			return execDoc(c, args, io)
+		},
+	)
+}
+
+func (c *docCfg) RegisterFlags(fs *flag.FlagSet) {
+	fs.BoolVar(
+		&c.all,
+		"all",
+		false,
+		"show documentation for all symbols in package",
+	)
+
+	fs.BoolVar(
+		&c.src,
+		"src",
+		false,
+		"show source code for symbols",
+	)
+
+	fs.BoolVar(
+		&c.unexported,
+		"u",
+		false,
+		"show unexported symbols as well as exported",
+	)
+
+	c.rootDirStruct.RegisterFlags(fs)
+}
+
+func execDoc(cfg *docCfg, args []string, io *commands.IO) error {
+	// guess opts.RootDir
+	if cfg.rootDir == "" {
+		cfg.rootDir = guessRootDir()
+	}
+	dirs := doc.NewDirs(filepath.Join(cfg.rootDir, "gnovm/stdlibs"), filepath.Join(cfg.rootDir, "examples"))
+	res, err := doc.ResolveDocumentable(dirs, args, cfg.unexported)
+	switch {
+	case res == nil:
+		return err
+	case err != nil:
+		io.Printfln("warning: error parsing some candidate packages:\n%v", err)
+	}
+	err = res.Document(
+		doc.WithShowAll(cfg.all),
+		doc.WithSource(cfg.src),
+		doc.WithUnexported(cfg.unexported),
+		doc.WithWriter(io.Out),
+	)
```

Comment:
What are the benefits of this approach in contrast to:

```suggestion
	err = res.Document{
		ShowAll:    cfg.all,
		Source:     cfg.src,
		Unexported: cfg.unexported,
		Writer:     io.Out,
	}
```

Reply:
Short answer: I like it better :smile: (It may or may not be that [ent's](https://entgo.io/) API design has made its way as a stable resident of my brain, but I digress)

Long answer: while in this case we're simply reporting to the doc package the options as given to the CLI, I think the `With*` pattern for optional function options works better because it stresses how arguments are optional for general users of the API, it still works very nicely in GoDoc, allows for default values which are not the zero value, and still looks pretty nice when using it.

You're right in saying a struct does the job just as well here and is still idiomatic Go, so I'm happy to go with whatever you find most readable.